### PR TITLE
Feature: File Atomic Write

### PR DESCRIPTION
**Pull request by chris-huxtable**
This PR adds `File.atomic_write()` to which prevents file corruption during a write.

### CODE REVIEW

File: src/file.cr [hunk 1]
```diff
@@ -672,6 +672,80 @@ class File < IO::FileDescriptor
...
+  # This is done by saving the new contents at temporary path. When the new content is
+  # successfully written the temporary path is changed to the provided path ensuring the data is
+  # not corrupted. If the write fails the temporary file is deleted.
+  def self.atomic_write(path : String, perm = DEFAULT_CREATE_PERMISSIONS, encoding = nil, invalid = nil, *, append : Bool = false, &block : IO::FileDescriptor -> Nil) : Bool
+    atomic_path = new_atomic_path(path)
+    success_flag = false
+
+    begin
+      open(atomic_path, "w", perm, encoding, invalid) do |fd|
+        fd.flock_exclusive() do
```
Comment (Sija) [inline src/file.cr]: Redundant `()`.

File: src/file.cr [hunk 2]
```diff
@@ -672,6 +672,80 @@ class File < IO::FileDescriptor
...
+          open(path, "r") { |src| IO.copy(src, fd) } if append
+          yield(fd)
+          fd.flush
+          success_flag = true
+        end
+      end
+    ensure
+      if (success_flag)
```
Comment (Sija) [inline src/file.cr]: Redundant `()`

File: src/file.cr [hunk 3]
```diff
@@ -672,6 +672,80 @@ class File < IO::FileDescriptor
...
+        atomic_install(atomic_path, path)
+      else
+        delete(atomic_path)
+      end
+    end
+
+    return success_flag
```
Comment (Sija) [inline src/file.cr]: Redundant `return`

File: src/file.cr [hunk 4]
```diff
@@ -672,6 +672,80 @@ class File < IO::FileDescriptor
...
+      when Bytes then fd.write(content)
+      when IO    then IO.copy(content, fd)
+      else            fd.print(content)
+      end
+    end
+  end
+
+  # :nodoc:
+  protected def self.atomic_install(atomic_path : String, dest_path : String) : Nil
+    if (info = info?(dest_path))
```
Comment (Sija) [inline src/file.cr]: Redundant `()`

File: src/file.cr [hunk 5]
```diff
@@ -672,6 +672,80 @@ class File < IO::FileDescriptor
...
+      chown(atomic_path, info.owner, info.group)
+    end
+    rename(atomic_path, dest_path)
+  end
+
+  # :nodoc:
+  protected def self.new_atomic_path(path : String, length : Int::Unsigned = 16_u8, limit : Int::Unsigned = 8_u8) : String
+    atomic_path = "#{path}.atomic_#{Random::Secure.urlsafe_base64(length)}"
+
+    while (exists?(atomic_path))
```
Comment (Sija) [inline src/file.cr]: Redundant `()`

File: src/file.cr [hunk 6]
```diff
@@ -672,6 +672,80 @@ class File < IO::FileDescriptor
...
+      raise "Failed to generate temporary path." if (limit <= 0)
+      atomic_path = "#{path}.atomic_#{Random::Secure.urlsafe_base64(length)}"
+      limit -= 1
+    end
+
+    return atomic_path
```
Comment (Sija) [inline src/file.cr]: Redundant `return`

File: src/file.cr (hunk 4 above)
Comment (chris-huxtable) [reply src/file.cr]: I didn't realize brackets were redundant on assignment. Interesting.

File: src/file.cr [hunk 7]
```diff
@@ -672,6 +672,80 @@ class File < IO::FileDescriptor
...
+    end
+    rename(atomic_path, dest_path)
+  end
+
+  # :nodoc:
+  protected def self.new_atomic_path(path : String, length : Int::Unsigned = 16_u8, limit : Int::Unsigned = 8_u8) : String
+    atomic_path = "#{path}.atomic_#{Random::Secure.urlsafe_base64(length)}"
+
+    while (exists?(atomic_path))
+      raise "Failed to generate temporary path." if (limit <= 0)
```
Comment (Sija) [inline src/file.cr]: Exception message IMO should mention the actual reason of the failure - without following dot `.`.
Redundant `()` too.

File: src/file.cr [hunk 8]
```diff
@@ -672,6 +672,80 @@ class File < IO::FileDescriptor
...
+  #
+  # NOTE: If the content is a `Slice(UInt8)`, those bytes will be written.
+  # If it's an `IO`, all bytes from the `IO` will be written.
+  # Otherwise, the string representation of *content* will be written
+  # (the result of invoking `to_s` on *content*).
+  def self.atomic_write(path : String, content, perm = DEFAULT_CREATE_PERMISSIONS, encoding = nil, invalid = nil) : Bool
+    atomic_write(path, perm, encoding, invalid) do |fd|
+      case content
+      when Bytes then fd.write(content)
+      when IO    then IO.copy(content, fd)
```
Comment (oprypin) [inline src/file.cr]: Consider making different method overloads

Comment (chris-huxtable) [reply src/file.cr]: I just copied `File.write()` which uses this method. I suspect its for verbosity reasons.

File: src/file.cr (hunk 7 above)
Comment (oprypin) [reply src/file.cr]: `raise "string"` won't do, needs to be an exception class

File: src/file.cr [hunk 9]
```diff
@@ -672,6 +672,80 @@ class File < IO::FileDescriptor
...
+      open(atomic_path, "w", perm, encoding, invalid) do |fd|
+        fd.flock_exclusive do
+          open(path, "r") { |src| IO.copy(src, fd) } if append
+          yield(fd)
+          fd.flush
+          success_flag = true
+        end
+      end
+    ensure
+      if success_flag
```
Comment (RX14) [inline src/file.cr]:
```cr
open(...) do |fd|
  # ...
  atomic_install(atomic_path, path)
rescue ex
  delete(atomic_path)
  raise ex
end
```

should work?

File: src/file.cr [hunk 10]
```diff
@@ -672,6 +672,80 @@ class File < IO::FileDescriptor
...
+        atomic_install(atomic_path, path)
+      else
+        delete(atomic_path)
+      end
+    end
+
+    success_flag
```
Comment (RX14) [inline src/file.cr]: This shouldn't return a bool: it always returns true or raises an exception.

File: src/file.cr [hunk 11]
```diff
@@ -672,6 +672,80 @@ class File < IO::FileDescriptor
...
+      chown(atomic_path, info.owner, info.group)
+    end
+    rename(atomic_path, dest_path)
+  end
+
+  # :nodoc:
+  protected def self.new_atomic_path(path : String, length : Int::Unsigned = 16_u8, limit : Int::Unsigned = 8_u8) : String
+    atomic_path = "#{path}.atomic_#{Random::Secure.urlsafe_base64(length)}"
+
+    while exists?(atomic_path)
```
Comment (RX14) [inline src/file.cr]: Just hardcode `length = 16`, then you're using 128bits of random and can assume there's no clashes. Then you can just remove this method.

File: src/file.cr [hunk 12]
```diff
@@ -682,28 +682,26 @@ class File < IO::FileDescriptor
...
+      delete(atomic_path)
+      raise ex
     end
 
-    success_flag
+    if info = info?(path)
+      chmod(atomic_path, info.permissions)
+      chown(atomic_path, info.owner, info.group)
+    end
+    rename(atomic_path, path)
```
[the remark this answers is not in the document]
Comment (bew) [reply src/file.cr]: I don't think so, chmod/chown/rename should be done only if the write was successful (no exception)

**Comment by asterite** [comment]
This should be in a shard.

File: src/file.cr (hunk 12 above)
[the remark this answers is not in the document]
Comment (bew) [reply src/file.cr]: You mean, in case chmod/chown fails?

**Comment by Sija** [comment]
@asterite Why do you think so?

**Comment by RX14** [comment]
I largely agree. This isn't an OS primitive, so I think it's a bit too opinionated and special-case to be in the stdlib.

**Comment by oprypin** [comment]
So, @chris-huxtable, don't take this as a dismissal. Please do actually make a shard, it would be very useful :)

**Comment by asterite** [comment]
Indeed! For example there's [this](https://github.com/dchest/safefile) for Go. There's [this](https://apidock.com/rails/File/atomic_write/class) (and I think others) for Ruby, etc.

**Comment by chris-huxtable** [comment]
I have written a [shard](https://github.com/chris-huxtable/atomic_write.cr) and updated it to what is in this PR. If any one else is interested.

**Comment by oprypin** [comment]
Of course, it would be better to not reopen a standard library module there

**Comment by chris-huxtable** [comment]
One of the reasons I think it should be in the stdlib (not to mention I think atomic writes should probably be the default). This is similar to [atomic_write](https://apidock.com/rails/File/atomic_write/class), referenced above. I also don't prescribe to the philosophy which promotes the prohibition of class extension. I come from Objective-C which treats this as a positive feature of a language. That being said, I realize it can be improperly maintained. This also isn't really a good place to have this argument.

**Comment by RX14** [comment]
Should we close then?

**Pull request closed without merging**